### PR TITLE
Fix CHPL_LLVM=system for Mac Mojave users using llvm@11 from homebrew

### PR DESCRIPTION
**Pull request by bradcray**
When using llvm@11 from homebrew on Mac Mojave, and using the
version of clang/clang++ it provides with the system linker,
the error message:

```
ld: unknown option: -platform_version
```

was being produced, as has been noticed by several other projects
online.  A workaround for this is to tell clang to assume a specific
linker version via:

```
-mlinker-version=450
```

this PR does that by adding a check to `gather-clang-sysroot-arguments`
that checks the Mac OS version, and adds this option if it's 1.14.x,
permitting the build and compilations to pass.

Resolves https://github.com/Cray/chapel-private/issues/2069

### CODE REVIEW

**Comment by bradcray** [comment]
@mppf: This should resolve the llvm@11 on Mojave CHPL_LLVM=system builds.  Could you verify that it doesn't break / change behavior on your Mac (where it should be sufficient to check that running `util/config/gather-clang-sysroot-arguments` on your Mac on this branch vs. main generates the same result).

**Review by mppf — APPROVED** [review]
Yes, `make` still works on my 10.15 Mac and I do not see the extra argument in `./util/config/gather-clang-sysroot-arguments`. Thanks for figuring out a solution!

**Comment by bradcray** [comment]
@ben-albrecht : I'm looking for confirmation that this PR fixes your CHPL_LLVM=system builds on Mojave before merging (I think you kicked off a test of it last Friday).

**Comment by ben-albrecht** [comment]
>  (I think you kicked off a test of it last Friday).

Updating to llvm@11 failed on my machine due to a python issue (maybe pyenv causing trouble?). I'll need to debug that before actually building.

**Comment by bradcray** [comment]
Don't forget that `brew doctor` is your friend.

**Comment by ben-albrecht** [comment]
@bradcray - I reproduced the former error and confirmed this branch works for me on Mojave. I think this PR is good to merge as is.

However, I want to point out a separate issue I discovered with the clang that comes with homebrew-installed llvm on Mojave. Essentially, when installing and linking llvm@11 via homebrew on Mojave, `clang` fails to compile anything without the `mlinker-version=450` flag. 

To reproduce:

```sh
>sw_vers -productVersion
10.14.6
>brew install llvm@11
>brew link llvm@11 # this puts clang that comes with llvm@11 in PATH
>clang hello.c # fails
ld: unknown option: -platform_version
>clang hello.c -mlinker-version=450 # works
```

Because brew-installed clang cannot compile a C program, it obviously fails to compile Chapel as well (the `-mlinker-version=450` flag would need to be added to every `clang` call in `make` to fix this). 

Chances are someone else will run into this issue down the road, so we might consider adding something to address it, either (1) detecting the error and pointing the user to some documentation on how to fix the issue or (2) detecting the MacOS / clang combination and always throwing `-mlinker-version=450` during compilation. 

(1) is probably more reasonable because it's a specific workaround to fix an issue in homebrew's installation and it will help that user fix their clang installation in general rather than make it work only for Chapel.

**Comment by bradcray** [comment]
Thanks for testing it out Ben.  I'll probably wait until after CHIUW to merge this just because I won't be able to respond to any issues that it causes if it does.

I don't disagree with what you're saying about putting in more effort to make this more bulletproof, but I'm also not particularly worried about it at present.  Specifically:

* If the user has set up llvm's clang as their default compiler rather than Apple's clang (which is something we don't require), they've hopefully also got it in a mode where it's working for them (i.e., they can compile a `hello.c`) or else they won't be able to compile much of anything on their Mac.  If they haven't taken pains to have llvm's clang be their default (the `brew link` command, essentially), then I think they'll get Apple's when building the compiler, and everything will be OK.
* w.r.t. your point (1), it's not really clear to me that there _is_ any way for them to fix the issue in homebrew's installation short of upgrading away from Mojave.  From all the web-searching I did, this is the only workaround I found, and homebrew doesn't seem interested in fixing the issue (maybe not interested in supporting Mojave at all, though I'm not sure which is the chicken and which is the egg).

I'm not opposed to having our Makefiles always use `-mlinker-version` for this case, but am inclined to put it off until it becomes an actual issue for someone.

Just to clarify one thing:  I feel fairly certain that the `-mlinker_version` flag is only required on the `clang` invocation that does a link step—so, not on any of the `clang -c` lines, which constitute most of our invocations.  So I'd think of this as being an LDFLAGS type of setting rather than a CFLAGS type of setting.

**Comment by ben-albrecht** [comment]
Yeah, I guess I left out option (0) of doing nothing because it's a user error, which I am also OK with until/unless this issue gains more attention from users encountering it.

> w.r.t. your point (1), it's not really clear to me that there is any way for them to fix the issue in homebrew's installation short of upgrading away from Mojave. From all the web-searching I did, this is the only workaround I found, and homebrew doesn't seem interested in fixing the issue (maybe not interested in supporting Mojave at all, though I'm not sure which is the chicken and which is the egg).

In my case, the fix was simply `brew unlink llvm@11`, but I don't know how generally that can be applied.

> Just to clarify one thing: I feel fairly certain that the -mlinker_version flag is only required on the clang invocation that does a link step—so, not on any of the clang -c lines, which constitute most of our invocations. So I'd think of this as being an LDFLAGS type of setting rather than a CFLAGS type of setting.

Agreed.

**Comment by bradcray** [comment]
> the fix was simply brew unlink llvm@11,

I see, I thought you were talking about a true fix to make it work rather than an environment fix to back away from using it.

File: util/config/gather-clang-sysroot-arguments
Comment (mppf) [inline util/config/gather-clang-sysroot-arguments:32]: I think maybe you meant to use `==` here? I'm seeing errors on a SLES 12 system about `sw_vers` not being found.

Comment (bradcray) [reply util/config/gather-clang-sysroot-arguments:32]: OK, thanks for reporting.  I think you're probably right (though I thought I'd just copied this off the internet somewhere and had to google `eq` vs. `==` to know what the difference is myself).  Just out of curiosity, what does `$CHPL_HOME/util/chplenv/chpl_platform.py --target` return on that system?

Comment (mppf) [reply util/config/gather-clang-sysroot-arguments:32]:
```
$ ./util/chplenv/chpl_platform.py --target
linux64
```

Comment (bradcray) [reply util/config/gather-clang-sysroot-arguments:32]: That's what I was guessing...  I wonder why we didn't have far more explosions in testing...  Is the definition of `-eq` on non-numeric strings undefined?  Or maybe it changed somehow uniquely in SLES 12?

In any case, I agree that `==` looks like what I should've used, and opened #17876 to address.

**Pull request merged**
